src/parser/inlines.rs

```diff
@@ -29,6 +29,10 @@ pub struct Subject<'a: 'd, 'r, 'o, 'd, 'i> {
     special_chars: [bool; 256],
     skip_chars: [bool; 256],
     smart_chars: [bool; 256],
+    // Need to borrow the callback from the parser only for the lifetime of the Subject, 'subj, and
+    // then give it back when the Subject goes out of scope. Needs to be a mutable reference so we
+    // can call the FnMut and let it mutate its captured variables.
+    callback: Option<&'subj mut &'c mut dyn FnMut(&[u8]) -> Option<(Vec<u8>, Vec<u8>)>>,
```

Comment:
![jaw drop](https://media.giphy.com/media/Nm8ZPAGOwZUQM/giphy.gif)

Reply:
Heh. I totally agree that this super ugly! I spent a long time staring at it trying to figure out if it could be made nicer. Unfortunately, it can't because of two things:

1. we need a reference to the closure because you can only have one instance of a mutable reference at any given time. You can futher borrow that reference (as we have done here), and the only other alternative is moving the reference around everywhere which is far more ugly!
2. you can only use a *mutable* reference through another *mutable* reference. You can't use it through an *immutable* reference. So `&'a &'b mut` would not work.

It's ugly, but it gives us memory safety so I guess it's okay! :sob: :laughing: